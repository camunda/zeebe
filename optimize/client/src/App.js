/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {HashRouter as Router, Route, Switch, matchPath} from 'react-router-dom';

import {
  PrivateRoute,
  Home,
  Collection,
  Report,
  Dashboard,
  Analysis,
  Sharing,
  Logout,
  Processes,
  ProcessReport,
} from './components';

import {ErrorBoundary, ErrorPage, Button} from 'components';

import {Notifications} from 'notifications';
import {SaveGuard} from 'saveGuard';
import {Prompt} from 'prompt';
import {Tracking} from 'tracking';
import {Onboarding} from 'onboarding';
import {ConfigProvider} from 'config';
import {Provider as Theme} from 'theme';
import {UserProvider, DocsProvider} from 'HOC';
import {TranslationProvider} from 'translation';

export default function App({error}) {
  function renderEntity(props) {
    const components = {
      report: Report,
      'dashboard/instant': Dashboard,
      dashboard: Dashboard,
      'processes/report': ProcessReport,
      collection: Collection,
    };
    const entities = ['processes/report', 'report', 'dashboard/instant', 'dashboard', 'collection'];
    let Component, newProps, selectedEntity;
    for (let entity of entities) {
      const splitResult = props.location.pathname.split('/' + entity)[1];
      if (splitResult) {
        const match = matchPath(splitResult, {path: '/:id/:viewMode?'});
        newProps = {
          ...props,
          match,
          entity,
        };
        Component = components[entity];
        selectedEntity = entity;
        break;
      }
    }
    // we do this to get a fresh component whenever the entity type changes
    // this is needed to be able to create new dashboard from instant preview dashboard page
    return <Component key={selectedEntity} {...newProps} />;
  }

  if (error) {
    return (
      <ErrorPage noLink text="Optimize could not be loaded, please make sure the server is running">
        <Button link onClick={() => window.location.reload(true)}>
          Reload
        </Button>
      </ErrorPage>
    );
  }

  return (
    <ConfigProvider>
      <TranslationProvider>
        <Theme>
          <Router getUserConfirmation={SaveGuard.getUserConfirmation}>
            <div className="Root-container">
              <ErrorBoundary>
                <UserProvider>
                  <DocsProvider>
                    <Switch>
                      <PrivateRoute exact path="/" component={Processes} />
                      <PrivateRoute path="/analysis" component={Analysis} />
                      <Route exact path="/share/:type/:id" component={Sharing} />
                      <PrivateRoute
                        path="/(report|dashboard/instant|dashboard|collection|processes/report)/*"
                        render={renderEntity}
                      />
                      <PrivateRoute exact path="/collections" component={Home} />
                      <Route path="/logout" component={Logout} />
                      <PrivateRoute path="*" component={ErrorPage} />
                    </Switch>
                  </DocsProvider>
                  <Tracking />
                  <Onboarding />
                </UserProvider>
              </ErrorBoundary>
            </div>
            <SaveGuard />
            <Prompt />
          </Router>
          <Notifications />
        </Theme>
      </TranslationProvider>
    </ConfigProvider>
  );
}
