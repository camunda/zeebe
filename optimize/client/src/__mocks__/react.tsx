/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

// @ts-ignore
export * from 'react';
export {default} from 'react';

const outstandingEffects: (() => void)[] = [];
const outstandingCleanups: (() => void)[] = [];

export const useEffect = (fn: () => (() => void) | void) => outstandingEffects.push(fn);
export const runLastEffect = () => {
  if (outstandingEffects.length) {
    const cleanup = outstandingEffects.pop()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

export const runLastCleanup = () => {
  if (outstandingCleanups.length) {
    outstandingCleanups.pop()?.();
  }
};

export const runAllEffects = () => {
  const numberOfEffects = outstandingEffects.length;

  for (let i = 0; i < numberOfEffects; i++) {
    const cleanup = outstandingEffects.shift()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

beforeEach(() => {
  outstandingEffects.length = 0;
});

export const runAllCleanups = () => {
  const numberOfEffects = outstandingCleanups.length;

  for (let i = 0; i < numberOfEffects; i++) {
    outstandingCleanups.shift()?.();
  }
};
