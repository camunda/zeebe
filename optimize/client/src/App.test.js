/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {runAllEffects} from 'react';
import {shallow} from 'enzyme';

import App from './App';

jest.mock('notifications', () => ({addNotification: jest.fn(), Notifications: () => <span />}));

jest.mock('hooks', () => ({
  useErrorHandling: () => ({
    mightFail: jest.fn().mockImplementation((data, cb) => cb(data)),
  }),
  useUser: jest.fn(),
}));

it('should show an error message when it is not possible to initilize the translation', async () => {
  const node = shallow(<App error="test error message" />);

  expect(node).toMatchSnapshot();
});

it('should render the last component in the url', () => {
  const node = shallow(<App />);

  runAllEffects();

  const component = node
    .find({
      path: '/(report|dashboard/instant|dashboard|collection|processes/report)/*',
    })
    .prop('render')({
    location: {pathname: '/collection/cid/dashboard/did/report/rid'},
  });
  const renderedEntity = shallow(component);

  expect(renderedEntity.dive().name()).toBe('Report');
  expect(renderedEntity.props().match.params.id).toBe('rid');
});

it('should render dashboar for dashboard/instant route', () => {
  const node = shallow(<App />);

  runAllEffects();

  const component = node
    .find({
      path: '/(report|dashboard/instant|dashboard|collection|processes/report)/*',
    })
    .prop('render')({
    location: {pathname: '/dashboard/instant/defKey'},
  });
  const renderedEntity = shallow(component);

  expect(renderedEntity.dive().name()).toBe('Dashboard');
  expect(renderedEntity.props().match.params.id).toBe('defKey');
});
