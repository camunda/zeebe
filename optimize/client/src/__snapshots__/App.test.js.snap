// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show an error message when it is not possible to initilize the translation 1`] = `
<ErrorPage
  noLink={true}
  text="Optimize could not be loaded, please make sure the server is running"
>
  <ForwardRef(Button)
    link={true}
    onClick={[Function]}
  >
    Reload
  </ForwardRef(Button)>
</ErrorPage>
`;
